### DISCUSSION TITLE
microdot utemplate difficulty

### DISCUSSION THREAD

**Discussion by CarlSchneider** [miguelgrinberg/microdot#247] [General]
Hi there.

I'm having difficulty getting microdot utemplate working on a Pico W running Micropython and hoping somebody can please point me in the right direction.

```python
from microdot import Microdot, send_file, Response # works expected and I can serve files with microdot.
```

In the examples for utilising utemplate the following is reflected

```python
from microdot.utemplate import Template # this just does not work for me. Errors with
Traceback (most recent call last):
  File "<stdin>", line 12, in <module>
ImportError: no module named 'microdot.utemplate'
```

So then I tried

```python
from utemplate import Template #thinking its a directory thing but this also errors out with
Traceback (most recent call last):
  File "<stdin>", line 14, in <module>
ImportError: no module named 'utemplate.Template'
``
However this does not error

```python
import utemplate # but I can't seem to get any template rendering to work. Even the 
Template.initialize('/')
``
errors with 
```python
  File "<stdin>", line 20, in <module>
AttributeError: 'module' object has no attribute 'initialize'
```
and `utemplate.initialize('/')`

errors with
```python
  File "<stdin>", line 20, in <module>
NameError: name 'Template' isn't defined
```

My directory on the Pico W looks like this
```
/> utemplate #utemplate directory with compile, recompile and source
```
the balance are all in the root directory
boot.py, main.py, microdot.py, template.py and sett.html

Using microdot I can `send_file(request.path)` fine but I just can't figure out how to render a html file using microdot with utemplate.

I had a look inside utemplate.py and I can see a class Template: but I'm stumped.

It must be something simple but just can't see it.

Hopefully its something really simple.

Any help would be appreciated.
Many thanks
Carl

**Comment by miguelgrinberg** [comment]
Microdot has a core file (microdot.py) and a few extensions. It seems you haven't done a correct installation.

On your device, create a `microdot` folder. Copy all the Python files from the `src/microdot` folder in this repository there. Then install the utemplate library (which is not part of this repository, it is a separate project maintained by a different person).

**Reply by CarlSchneider** [reply]
Thank you very much for your quick reply Miguel.

I have used Thonny Manage Packages to install microdot and utemplate.

The Pico W FS now looks like this
<img width="224" alt="image" src="https://github.com/user-attachments/assets/1d9c161d-beba-4757-ba9f-353582b2ff83">

And the template rendering seems to be doing something now-not the result I was expecting though :) so back to trying to figure out how to use utemplate properly.

Many thanks
Carl

**Reply by CarlSchneider** [reply]
Thanks again for the response and assistance.

utemplate now working as expected; was caught out by the default content type for microdot being text/plain but once changed to text/html it all worked.

Live and learn :)
